docs/modules/SDDs/pages/0021-cluster-catalog-compilation.adoc

```diff
@@ -0,0 +1,170 @@
+
+Quartermaster can use reclass to parse the inventory and search for clusters that are affected by changes in a certain class.
+One issue we've to work around in this approach is, that reclass can only parse the inventory if it's complete and all referenced classes are available.
+This is a problem since creating the whole inventory (with all components) would be too resource and time intensive.
+As a workaround Quartermaster should only have to get the global and tenant parts of the inventory.
+The classes for the different Commodore components should be faked and generated as empty files.
+The only downside to this approach is that if a component class imports another class, this won't be captured by Quartermaster which might result in cluster catalogs not being compiled.
+Since we currently don't allow components to include classes this isn't a problem.
+
+
+==== Webhooks
+
+Triggering a webhook can be implemented in either a Git repo natively (for example https://developer.github.com/webhooks/[GitHub Webhooks]) or via a CI/CD system (for example a https://docs.gitlab.com/ee/ci/[GitLab CI] job which runs `curl`).
+A webhook can also be called by any other tool or even manually with a `curl` command.
+
+A webhook can include the following optional information:
+
+* Cluster ID for which to compile the catalog.
+This can be used in case where a specific cluster should be compiled
+
+* Tenant ID for which to figure out affected clusters.
+This can be used to limit the set of clusters to be considered
+
+* Set of classes that changed
+
+
+==== Authentication & Authorization
+
+To prevent exsessive resource usage or even DoS attacks, the webhooks need to be authenticated.
+A permission system needs to be in place to decide which entities are allowed to trigger which actions.
+It shouldn't be possible for example for a tenant to trigger compilation for clusters of other tenants.
+
+
+=== Compilation Jobs
+
+To compile the catalog for a cluster, a compilation job is created.
+The compilation job is a https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion[Kubernetes job] resource which executes the xref:0006-commodore-configuration-catalog-compilation.adoc[catalog compilation engine].
+
+To distribute the load this compilation generates and to follow a decentralized approach, these jobs are run on the same cluster as they compile the catalog for.
+The Steward component on a cluster starts a compilation job if the `cluster.syn.tools/needs-compilation` annotation is set to `true` on its own cluster object.
+It will then launch the respective job to compile the cluster catalog.
+
+For clusters where this isn't possible due to resource, security or other constraints, a central system runs the jobs.
+These clusters need to be configured in the central system so it will react on the mentioned annotation an create the compile jobs for them.
```

Comment:
So will it really be Steward creating the job objects? At least the code we write to do so needs to be reusable so it can be integrated in Steward but also be used on surrogate systems hosting the compilation for another cluster.

I imagine this to be an project on APPUiO per tenant with surrogate catalogue compilation needs.

Reply:
I added a note to make sure the logic when/how the job is run is the same on clusters and the central system.
How exactly this will look like can be defined at a later point.